lib/checkstl.cpp

```diff
@@ -3117,6 +3117,52 @@ void CheckStl::knownEmptyContainer()
     }
 }
 
+void CheckStl::eraseEndIteratorError(const Token *ftok, const Token* itertok)
+{
+    const std::string func = ftok ? ftok->str() : std::string("erase");
+    const std::string iter = itertok ? itertok->expressionString() : std::string("it");
+
+    const std::string msg = "Function '" + func + "()' must not be called on the end iterator '" + iter + "'.";
+
+    reportError(ftok, Severity::error,
+                "eraseEndIterator",
+                msg, CWE398, Certainty::normal);
+}
+
+void CheckStl::eraseEndIterator()
+{
+    logChecker("CheckStl::eraseEndIterator");
+    for (const Scope *function : mTokenizer->getSymbolDatabase()->functionScopes) {
+        for (const Token *tok = function->bodyStart; tok != function->bodyEnd; tok = tok->next()) {
+
+            if (!tok->valueType())
+                continue;
+            const Library::Container* container = tok->valueType()->container;
+            if (!container)
+                continue;
+            if (!Token::Match(tok->next(), ". %name% ("))
+                continue;
+            const Library::Container::Action action = container->getAction(tok->strAt(2));
+            if (action != Library::Container::Action::ERASE)
+                continue;
+            const std::vector<const Token*> args = getArguments(tok->tokAt(2));
+            if (args.size() != 1) // empty range is ok
+                continue;
+
+            if (const ValueFlow::Value* endVal = args[0]->getKnownValue(ValueFlow::Value::ValueType::ITERATOR_END)) {
+                if (endVal->intvalue == 0)
+                    eraseEndIteratorError(tok->tokAt(2), args[0]);
+            }
+            else if (const ValueFlow::Value* startVal = args[0]->getKnownValue(ValueFlow::Value::ValueType::ITERATOR_START)) {
+                if (startVal->intvalue == 0)
```

Comment:
Well we should check the iterator is out of bounds not just begin on an empty vector: `startVal->intvalue >= sizeVal->intvalue`